contracts/chain-adapters/interfaces/ArbitrumForwarderInterface.sol

```diff
@@ -0,0 +1,254 @@
+    // Gateway used for communicating with chains that use custom gas tokens:
+    // https://github.com/OffchainLabs/token-bridge-contracts/blob/main/contracts/tokenbridge/ethereum/gateway/L1ERC20Gateway.sol
+    ArbitrumERC20GatewayLike public immutable L2_ERC20_GATEWAY_ROUTER;
+
+    error RescueFailed();
+
+    /*
+     * @dev All functions with this modifier must revert if msg.sender != CROSS_DOMAIN_ADMIN, but each L2 may have
+     * unique aliasing logic, so it is up to the forwarder contract to verify that the sender is valid.
+     */
+    modifier onlyAdmin() {
+        _requireAdminSender();
+        _;
+    }
+
+    /**
+     * @notice Constructs new Adapter.
+     * @param _l2ArbitrumInbox Inbox helper contract to send messages to Arbitrum-like L3s.
+     * @param _l2ERC20GatewayRouter ERC20 gateway router contract to send tokens to Arbitrum-like L3s.
+     * @param _l3RefundL3Address L3 address to receive gas refunds on after a message is relayed.
+     * @param _l3MaxSubmissionCost Amount of gas token allocated to pay for the base submission fee. The base
+     * submission fee is a parameter unique to Arbitrum retryable transactions. This value is hardcoded
+     * and used for all messages sent by this adapter.
+     * @param _l3SpokePool L3 address of the contract which will receive messages and tokens which are temporarily
+     * stored in this contract on L2.
+     * @param _crossDomainAdmin L1 address of the contract which can send root bundles/messages to this forwarder contract.
+     * In practice, this is the hub pool.
+     */
+    constructor(
+        ArbitrumInboxLike _l2ArbitrumInbox,
+        ArbitrumERC20GatewayLike _l2ERC20GatewayRouter,
+        address _l3RefundL3Address,
+        uint256 _l3MaxSubmissionCost,
+        address _l3SpokePool,
+        address _crossDomainAdmin
+    ) {
+        L2_INBOX = _l2ArbitrumInbox;
+        L2_ERC20_GATEWAY_ROUTER = _l2ERC20GatewayRouter;
+        L3_REFUND_L3_ADDRESS = _l3RefundL3Address;
+        L3_MAX_SUBMISSION_COST = _l3MaxSubmissionCost;
+        L3_SPOKE_POOL = _l3SpokePool;
+        CROSS_DOMAIN_ADMIN = _crossDomainAdmin;
+    }
+
+    /**
+     * @notice When called by the cross domain admin (i.e. the hub pool), the msg.data should be some function
+     * recognizable by the L3 spoke pool, such as "relayRootBundle" or "upgradeTo". Therefore, we simply forward
+     * this message to the L3 spoke pool using the implemented messaging logic of the L2 forwarder
+     */
+    fallback() external onlyAdmin {
+        _relayMessage(L3_SPOKE_POOL, msg.data);
+    }
+
+    /**
+     * @notice This function can only be called via a rescue adapter, and is used to recover potentially stuck
+     * funds on this contract.
+     */
+    function rescue(
+        address target,
+        uint256 value,
+        bytes memory message
+    ) external onlyAdmin {
+        (bool success, ) = target.call{ value: value }(message);
```

Comment:
There are a bunch of ways to rescue tokens. Should this be kept like this, or should we delegatecall to a target, or should we just withdraw the tokens back to the specified address, or something else?

Reply:
The function, as built, can do far more than rescue tokens. It can perform any action on behalf of the contract. We should probably be explicit about that in the naming. Something more like `adminCall` or similar.

Relatedly: there are reasons we might want a function that gives the admin broad power like this. For instance, if the L2->L3 canonical bridge interface changes underneath us and this contract is non-upgradable (something we should discuss), it could brick the L3 SpokePool without a function like this that can be used as an escape-hatch to make an arbitrary call.

We should decide whether we want the escape hatch to be upgradability OR an arbitrary call admin method like this.

Thoughts?

Reply:
Made into a proxy here: https://github.com/across-protocol/contracts/pull/609/commits/f2e3c41f3a0f78ba94d2ee2dc749a782e54512e0